src/client/microsoftAuth.js

```diff
@@ -52,21 +72,132 @@ module.exports = async (client, options) => {
   }
 
   const session = {
-    accessToken: MineServicesResponse.access_token,
+    accessToken: mcAccessToken,
     selectedProfile: profile,
     availableProfile: [profile]
   }
   client.session = session
   client.username = MinecraftProfile.name
-  options.accessToken = MineServicesResponse.access_token
+  options.accessToken = authConstants.MineServicesResponse.access_token
   client.emit('session', session)
   options.connect(client)
 }
 
+/**
+ * Authenticates with Mincrosoft through user credentials, then
+ * with Xbox Live, Minecraft, checks entitlements and returns profile
+ *
+ * @function
+ * @param {object} client - The client passed to protocol
+ * @param {object} options - Client Options
+ */
+async function authenticatePassword (client, options) {
+  // Use external library to authenticate with
+
+  const XAuthResponse = await XboxLiveAuth.authenticate(options.username, options.password, { XSTSRelyingParty: authConstants.XSTSRelyingParty })
+    .catch((err) => {
+      console.warn('Unable to authenticate with Microsoft', err)
+      throw err
+    })
+
+  const MineServicesResponse = await fetch(authConstants.MinecraftServicesLogWithXbox, {
+    method: 'post',
+    ...getFetchOptions,
+    body: JSON.stringify({ identityToken: `XBL3.0 x=${XAuthResponse.userHash};${XAuthResponse.XSTSToken}` })
+  }).then(checkStatus)
+
+  getFetchOptions.headers.Authorization = `Bearer ${MineServicesResponse.access_token}`
+  const MineEntitlements = await fetch(authConstants.MinecraftServicesEntitlement, getFetchOptions).then(checkStatus)
+  if (MineEntitlements.items.length === 0) throw Error('This user does not have any items on its accounts according to minecraft services.')
+
+  postAuthenticate(client, options, MineServicesResponse.access_token)
+}
+
+async function getMsaToken () {
+  if (await msa.verifyTokens()) {
+    return msa.getAccessToken().token
+  } else {
+    const ret = await msa.authDeviceToken((response) => {
```

Comment:
authDeviceToken always returns nothing (in console)
![](https://uwu.whats-th.is/6xs4dT1.png)
regardless whether this is our first time authenticating or not.

Reply:
The device_code line is debug logging, the path issues I think were caused by not calling path.join, it should be fixed now